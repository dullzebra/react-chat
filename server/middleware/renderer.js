import React from 'react';
import ReactDOMServer from 'react-dom/server';
import { StaticRouter } from 'react-router-dom';
import { Provider } from 'react-redux';

import { SheetsRegistry } from 'react-jss/lib/jss';
import JssProvider from 'react-jss/lib/JssProvider';
import { createGenerateClassName } from 'material-ui/styles';

import configureStore from '../../src/store';
import App from '../../src/components/App';

const sheetsRegistry = new SheetsRegistry();
const generateClassName = createGenerateClassName();

const path = require('path');
const fs = require('fs');

export default (req, res) => {
  // point to the html file created by CRA's build tool
  const filePath = path.resolve(__dirname, '..', '..', 'build', 'index.html');
  const context = {};
  const store = configureStore();

  fs.readFile(filePath, 'utf8', (err, htmlData) => {
    if (err) {
      console.error('err', err);
      return res.status(404).end();
    }

    if (context.url) {
      res.redirect(context.url);
      return;
    }

    // render the app as a string
    /* eslint-disable-next-line */
    const html = ReactDOMServer.renderToString(
      <JssProvider registry={sheetsRegistry} generateClassName={generateClassName}>
        <Provider store={store}>
          <StaticRouter location={req.url} context={context}>
            <App />
          </StaticRouter>
        </Provider>
      </JssProvider>);

    const css = sheetsRegistry.toString();

    // inject the rendered app into our html and send it
    return res.end(htmlData.replace(
      '<div id="root"></div>',
      `<div id="root">${html}</div>
    <script>window.INITIAL_STATE = ${JSON.stringify(store.getState()).replace(
    /</g,
    '\\u003c',
  )}</script>
    <style id="jss-server-side">${css}</style>
    `,
    ));
  });
};
