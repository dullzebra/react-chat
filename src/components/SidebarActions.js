import React from 'react';
import PropTypes from 'prop-types';
import { withStyles } from 'material-ui/styles';
import BottomNavigation, { BottomNavigationAction } from 'material-ui/BottomNavigation';
import GroupIcon from 'material-ui-icons/Group';
import StarIcon from 'material-ui-icons/Star';
import AddIcon from 'material-ui-icons/Add';
import Button from 'material-ui/Button';
import AddChat from './AddChat';
import * as style from '../constants/styles';

const styles = theme => ({
  panel: {
    minHeight: style.bottomPanelHeight,
  },
  addChat: {
    position: 'absolute',
    bottom: style.bottomPanelHeight + theme.spacing.unit * 3,
    right: theme.spacing.unit * 3,
  },
  paper: {
    ...style.centerPosition,
    ...style.dialogWidth,
    backgroundColor: theme.palette.background.paper,
    boxShadow: theme.shadows[5],
    padding: theme.spacing.unit * 3,
  },
  footer: {
    display: 'flex',
    justifyContent: 'space-between',
    marginTop: theme.spacing.unit * 2,
  },
});

class SidebarActions extends React.Component {
  static propTypes = {
    classes: PropTypes.objectOf(PropTypes.string).isRequired,
    createChat: PropTypes.func.isRequired,
    showAllChats: PropTypes.func.isRequired,
    disabled: PropTypes.bool.isRequired,
  };

  state = {
    isModalOpen: false,
    navValue: 1,
  };

  handleNav = (event, navValue) => {
    this.setState({ navValue });
    this.props.showAllChats(!!navValue);
  };

  handleOpen = (e) => {
    e.preventDefault();
    this.setState({ isModalOpen: true });
  };

  handleClose = () => {
    this.setState({ isModalOpen: false });
  };

  render() {
    const { classes, disabled } = this.props;

    return (
      <React.Fragment>
        <BottomNavigation
          showLabels
          disabled={disabled}
          className={classes.panel}
          value={this.state.navValue}
          onChange={this.handleNav}
        >
          <BottomNavigationAction label="Мои чаты" icon={<StarIcon />} />
          <BottomNavigationAction label="Все чаты" icon={<GroupIcon />} />
        </BottomNavigation>

        <Button
          variant="fab"
          color="primary"
          className={classes.addChat}
          disabled={disabled}
          onClick={this.handleOpen}
        >
          <AddIcon />
        </Button>
        <AddChat
          open={this.state.isModalOpen}
          onClose={this.handleClose}
          onSubmit={this.props.createChat}
        />
      </React.Fragment>
    );
  }
}
export default withStyles(styles)(SidebarActions);
