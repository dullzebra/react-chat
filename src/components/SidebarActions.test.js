/* eslint-env jest */
import React from 'react';
import ReactDOM from 'react-dom';
import renderer from 'react-test-renderer';
import SidebarActions from './SidebarActions';

const mockData = {
  disabled: false,
  showAllChats: jest.fn(),
  createChat: jest.fn(),
};

describe('<Sidebar />', () => {
  it('renders without crashing', () => {
    const div = document.createElement('div');
    ReactDOM.render(<SidebarActions {...mockData} />, div);
    ReactDOM.unmountComponentAtNode(div);
  });

  it('renders disabled buttons', () => {
    const tree = renderer.create(<SidebarActions {...mockData} disabled />).toJSON();
    expect(tree).toMatchSnapshot();
  });
});
