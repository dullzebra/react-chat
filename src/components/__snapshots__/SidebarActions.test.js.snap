// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Sidebar /> renders disabled buttons 1`] = `
Array [
  <div
    className="MuiBottomNavigation-root-52 SidebarActions-panel-48"
    disabled={true}
>
    <button
        className="MuiButtonBase-root-60 MuiBottomNavigationAction-root-53"
        disabled={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
    >
        <span
            className="MuiBottomNavigationAction-wrapper-56"
        >
            <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-62"
                color={undefined}
                focusable="false"
                viewBox="0 0 24 24"
            >
                <path
                    d="M12 17.27L18.18 21l-1.64-7.03L22 9.24l-7.19-.61L12 2 9.19 8.63 2 9.24l5.46 4.73L5.82 21z"
                />
            </svg>
            <span
                className="MuiBottomNavigationAction-label-57"
            >
                Мои чаты
            </span>
        </span>
        <span
            className="MuiTouchRipple-root-69"
        />
    </button>
    <button
        className="MuiButtonBase-root-60 MuiBottomNavigationAction-root-53 MuiBottomNavigationAction-selected-54"
        disabled={undefined}
        onBlur={[Function]}
        onClick={[Function]}
        onFocus={[Function]}
        onKeyDown={[Function]}
        onKeyUp={[Function]}
        onMouseDown={[Function]}
        onMouseLeave={[Function]}
        onMouseUp={[Function]}
        onTouchEnd={[Function]}
        onTouchMove={[Function]}
        onTouchStart={[Function]}
        tabIndex={0}
        type="button"
    >
        <span
            className="MuiBottomNavigationAction-wrapper-56"
        >
            <svg
                aria-hidden="true"
                className="MuiSvgIcon-root-62"
                color={undefined}
                focusable="false"
                viewBox="0 0 24 24"
            >
                <path
                    d="M16 11c1.66 0 2.99-1.34 2.99-3S17.66 5 16 5c-1.66 0-3 1.34-3 3s1.34 3 3 3zm-8 0c1.66 0 2.99-1.34 2.99-3S9.66 5 8 5C6.34 5 5 6.34 5 8s1.34 3 3 3zm0 2c-2.33 0-7 1.17-7 3.5V19h14v-2.5c0-2.33-4.67-3.5-7-3.5zm8 0c-.29 0-.62.02-.97.05 1.16.84 1.97 1.97 1.97 3.45V19h6v-2.5c0-2.33-4.67-3.5-7-3.5z"
                />
            </svg>
            <span
                className="MuiBottomNavigationAction-label-57 MuiBottomNavigationAction-selectedLabel-58"
            >
                Все чаты
            </span>
        </span>
        <span
            className="MuiTouchRipple-root-69"
        />
    </button>
</div>,
  <button
    className="MuiButtonBase-root-60 MuiButtonBase-disabled-61 MuiButton-root-76 MuiButton-raised-81 MuiButton-fab-86 MuiButton-raisedPrimary-83 MuiButton-disabled-85 SidebarActions-addChat-49"
    disabled={true}
    onBlur={[Function]}
    onClick={[Function]}
    onFocus={[Function]}
    onKeyDown={[Function]}
    onKeyUp={[Function]}
    onMouseDown={[Function]}
    onMouseLeave={[Function]}
    onMouseUp={[Function]}
    onTouchEnd={[Function]}
    onTouchMove={[Function]}
    onTouchStart={[Function]}
    tabIndex={-1}
    type="button"
>
    <span
        className="MuiButton-label-77"
    >
        <svg
            aria-hidden="true"
            className="MuiSvgIcon-root-62 MuiSvgIcon-fontSize-68"
            color={undefined}
            focusable="false"
            viewBox="0 0 24 24"
        >
            <path
                d="M19 13h-6v6h-2v-6H5v-2h6V5h2v6h6v2z"
            />
        </svg>
    </span>
</button>,
]
`;
